### ISSUE TITLE
cmd/govim: error on SetUserBusy caused by bad buffer life-cycle events

### ISSUE THREAD

**Issue by ohait** [govim/govim#801]
### What version of Vim/Gvim are you using?

<details><summary>Vim/Gvim version</summary><br><pre>
$ vim --version
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Feb 06 2020 12:13:42)
Included patches: 1-221

</pre></details>

### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.13.1 linux/amd64

</pre>

### What operating system and processor architecture are you using (`go env`)?
ubuntu 18.4/amd64

### What version/commit of `govim` are you using?
latest via vimplug

### What is the output of `:messages` in Vim?
```
Error detected while processing function <SNR>13_userBusy[3]..GOVIM_internal_SetUserBusy[2]..<SNR>13_callbackFunction[2]..<SNR>13_ch_evalexpr:
line    4:
E605: Exception not caught: got error whilst handling GOVIM_internal_SetUserBusy: driver error: ChannelCall("s:batchCall") failed: failed to call s:batch
Call([[[call [s:mustNoError []] setqflist [] r]]]) in Vim: Caught 'failed to call setqflist([[], ''r'']): got error whilst handling autocommand:1: BufWin
Enter fired for buffer 4; but we don''t know about it' in function <SNR>13_userBusy[3]..GOVIM_internal_SetUserBusy[2]..<SNR>13_callbackFunction[2]..<SNR>
13_ch_evalexpr[2]..<SNR>13_define[61]..<SNR>13_batchCall, line 35
Press ENTER or type command to continue
Error detected while processing function <SNR>13_userBusy:
line    3:
E171: Missing :endif
Press ENTER or type command to continue
```

### Do the Vim channel, `govim` or `gopls` log files show anything interesting?
no errors or anything `funny`

### How did you install `govim`?
vimplug

### What did you do?
it usually happens after quickfix pops up after writing with an error

### What did you see instead?
some errors, which will then keep popping up

**Comment by myitcv** [comment]
Thank you for reporting this issue, @ohait 

Please see https://github.com/govim/govim/pull/802 for a likely fix.

Unfortunately we are going to have to work through the pain of these bugs on `master`. The "bug" was introduced when we rationalised the handling of buffer life-cycle events as part of a wider change to track all buffers in `govim`. These bugs appear to be because Vim is reporting strange out-of-sequence events (https://github.com/vim/vim/issues/5694, https://github.com/vim/vim/issues/5656, https://github.com/vim/vim/issues/5655)

**Comment by ohait** [comment]
thanks @myitcv !
after a PlugUpdate it got way way better!

still noticing this tho:
```
Error detected while processing function <SNR>13_userBusy[3]..GOVIM_internal_SetUserBusy[2]..<SNR>13_callbackFunction[2]..<SNR>13_ch_evalexpr:
line    4:
E605: Exception not caught: got error whilst handling GOVIM_internal_SetUserBusy: failed to get current position: buffer 2 is not loaded
Error detected while processing function <SNR>13_userBusy:
line    3:
E171: Missing :endif
```
and i believe it appears after a while on-idle. 

sorry i wish i knew more where to dig for this.

**Comment by myitcv** [comment]
Thanks @ohait - if you can provide some steps to repro that would be a great help. If you could also provide the `govim` log that would help too:

https://github.com/govim/govim/wiki/Debugging-govim-plugin-failures#log-files

Thanks

**Comment by myitcv** [comment]
FWIW @midnightrun provided a repro on Slack:

> * Clone: https://github.com/midnightrun/aggregator-pattern
> * Checkout the branch feature/handle_aggregation_processing
> * Enter the main directory and start Vim via vim
> * Open Netrw via :Explore and wait some seconds

But I'm not able to reproduce this, which suggests it's a config thing

@ohait - can you repro the problem using the above steps?

If so, can you share your `.vimrc`?

**Comment by ohait** [comment]
I can reproduce, and I can do better than giving you my `.vimrc`, i found what triggers it:
`set updatetime=700`

**Comment by myitcv** [comment]
@ohait - I have reproduced this using @midnightrun's `.vimrc`. It appears to be another Vim bug. We can work around this in `govim` - fix coming shortly.

**Comment by ohait** [comment]
by switching updatetime to the default of 4sec it works like a charm

**Comment by myitcv** [comment]
... unless you don't move the cursor for 4 seconds :)

**Comment by ohait** [comment]
oh, i thought i waited for 4 seconds, i'm too eager, sorry!

**Comment by a-palchikov** [comment]
Not sure how related my issue to this but it is definitely related to `setUserBusy`. I'm on https://github.com/govim/govim/commit/a3741ae96aa164e1e2bad5b920d724b49a5263cd which is the tip at this point.
This is the message that I consistently run into:
```
Exception not caught: got error whilst handling GOVIM_internal_SetUserBusy failed to get current position: failed to calculate offset within buffer 5: line is beyond end of file
```
when I'm in the temporary buffer with results from, say, `GOVIMReferences`.
`updatetime` is the default 4s, so it basically triggers **every** time I switch to the buffer or change the cursor position within that buffer and wait for `updatetime`.

**Comment by myitcv** [comment]
Thanks @a-palchikov - I'm working on a fix for this and other issues in #807 

For now the workaround is to stick with [`v0.0.29`](https://github.com/govim/govim/releases/tag/v0.0.29) or go back to b989e95684c9c6d30f7fc08f3a2f98c899153130

**Comment by myitcv** [comment]
Thanks everyone for your patience with this bug. I'm going to back out the changes that are causing this and other bugs. Per my update in https://github.com/vim/vim/issues/5694#issuecomment-594525215, we will have to look at different ways of having `govim` support non-`.go` buffers safely.
